components/hub/hub_service.bal

```diff
@@ -0,0 +1,244 @@
+    isolated remote function onSubscription(websubhub:Subscription message, http:Headers headers)
+                returns websubhub:SubscriptionAccepted|websubhub:BadSubscriptionError|error {
+        if config:securityOn {
+            check security:authorize(headers, ["subscribe"]);
+        }
+        return websubhub:SUBSCRIPTION_ACCEPTED;
+    }
+
+    # Validates a incomming subscription request.
+    #
+    # + message - Details of the subscription
+    # + return - `websubhub:SubscriptionDeniedError` if the subscription is denied by the hub or else `()`
+    isolated remote function onSubscriptionValidation(websubhub:Subscription message)
+                returns websubhub:SubscriptionDeniedError? {
+        boolean topicAvailable = false;
+        lock {
+            topicAvailable = registeredTopicsCache.hasKey(message.hubTopic);
+        }
+        if !topicAvailable {
+            return error websubhub:SubscriptionDeniedError(
+                "Topic [" + message.hubTopic + "] is not registered with the Hub", statusCode = http:STATUS_NOT_ACCEPTABLE);
+        } else {
+            string subscriberId = common:generateSubscriberId(message.hubTopic, message.hubCallback);
+            if isValidSubscription(subscriberId) {
+                return error websubhub:SubscriptionDeniedError(
+                    "Subscriber has already registered with the Hub", statusCode = http:STATUS_NOT_ACCEPTABLE);
+            }
+        }
+    }
+
+    # Processes a verified subscription request.
+    #
+    # + message - Details of the subscription
+    # + return - `error` if there is any unexpected error or else `()`
+    isolated remote function onSubscriptionIntentVerified(websubhub:VerifiedSubscription message) returns error? {
+        lock {
+            if !message.hasKey(CONSUMER_GROUP) {
+                string consumerGroup = common:generateGroupName(message.hubTopic, message.hubCallback);
+                message[CONSUMER_GROUP] = consumerGroup;
+            }
+            message[SERVER_ID] = config:serverIdentifier;
+            error? persistingResult = persist:addSubscription(message.cloneReadOnly());
+            if persistingResult is error {
+                log:printError("Error occurred while persisting the subscription ", err = persistingResult.message());
+            }
+        }
+    }
+
+    # Unsubscribes a `subscriber` from the hub.
+    #
+    # + message - Details of the unsubscription
+    # + headers - `http:Headers` of the original `http:Request`
+    # + return - `websubhub:UnsubscriptionAccepted` if unsubscription is accepted from the hub, `websubhub:BadUnsubscriptionError`
+    # if unsubscription is denied from the hub or `error` if there is any unexpected error
+    isolated remote function onUnsubscription(websubhub:Unsubscription message, http:Headers headers)
+                returns websubhub:UnsubscriptionAccepted|websubhub:BadUnsubscriptionError|error {
```

Comment:
Do we need the `error` as a return type here? Can't we use the exact error types returned?